### PR TITLE
doc(lxl): best practice - add buffer_zh

### PR DESCRIPTION
**Pull request by lixl-st**

### CODE REVIEW

File: source/04_best_practice/buffer_zh.rst [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    buffer = DequeBuffer(size=10)
+    buffer.use(clone_object())
+
+
+**分组采样**
+
+
+在某些算法或环境中，收集和存储整个剧集比分离样本更有用。例如：在国际象棋、围棋或纸牌游戏中，玩家只有在游戏结束时才能获得奖励。
+一些算法，例如 Hindsight Experience Replay (HER)，必须对一条条完整的 episode 进行采样并对其进行操作。
+这时，用户可以使用分组采样的方式实现这一目标。\ **分组采样会严重增加采样耗时**\。
```
Comment (puyuan1996) [inline source/04_best_practice/buffer_zh.rst:81]: 在存储样本 (buffer.push) 时，每次存入一整局样本， 同时通过设置 "episode" 的分组关键字， 即可以在采样样本时 (buffer.sample) 时指定 groupby="episode" 来实现以episode为单位的采样。

File: source/04_best_practice/buffer_zh.rst [hunk 2]
```diff
@@ -0,0 +1,104 @@
+Buffer 使用指南
+===============================
+
+Buffer 入门
+-------------------------------
+
+在 Off-policy 算法中，我们通常需要使用缓存池来实现经验回放（Experience Replay）机制。
```
Comment (puyuan1996) [inline source/04_best_practice/buffer_zh.rst:7]: 在 Off-policy RL 算法中，通常会使用经验回放（Experience Replay）机制来提高样本利用效率，为此需要定义一个缓存池来实现样本的存储，处理和采集。

Comment (Will-Nie) [reply source/04_best_practice/buffer_zh.rst:7]: First, each step of experience
is potentially used in many weight updates, which allows for greater data efficiency.
Second, learning directly from consecutive samples is inefficient, owing to the strong
correlations between the samples; randomizing the samples breaks these correlations
and therefore reduces the variance of the updates. Third, when learning on policy
the current parameters determine the next data sample that the parameters
are trained on. For example, if the maximizing action is to move left then the training
samples will be dominated by samples from the left-hand side; if the maximizing
action then switches to the right then the training distribution will also switch.
It is easy to see how unwanted feedback loops may arise and the parameters could get
stuck in a poor local minimum, or even diverge catastrophically20. By using experience 
replay the behaviour distribution is averaged over many of its previous states,
smoothing out learning and avoiding oscillations or divergence in the parameters.

补充下蒲源上面的提高样本效率（对应上面英文的第一点）， 也可以加一下第二点和第三点。

File: source/04_best_practice/buffer_zh.rst [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+接下来，用户可以通过 DI-engine 中封装好的中间件来调用该 buffer 对象，完成训练任务（\ **推荐方式**\）。
+
+.. code-block:: python
+
+    task.use(data_pusher(cfg, buffer))
+    task.use(OffPolicyLearner(cfg, policy.learn_mode, buffer))
+
+
+此外，当用户有其它使用 buffer 缓存数据的需求时，也可以通过自定义的方式实现数据的存储与采样，例如：
```
Comment (PaParaZz1) [inline source/04_best_practice/buffer_zh.rst:24]: 介绍下dataclass吧，要不然很多人一上来不理解为什么要这样，以及具体怎么用

File: source/04_best_practice/buffer_zh.rst [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+.. code-block:: python
+    
+    from ding.data.buffer.middleware import clone_object
+
+    buffer = DequeBuffer(size=10)
+    buffer.use(clone_object())
+
+
+**分组采样**
```
Comment (PaParaZz1) [inline source/04_best_practice/buffer_zh.rst:78]: 分组采样介绍一个具体的例子吧，以r2d2为例，可以看看这个入口相关的[实现](https://github.com/opendilab/DI-engine/blob/main/ding/example/r2d2.py)

File: source/04_best_practice/buffer_zh.rst [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+(可选)另外，在分组采样的基础上，还通过 group_sample 中间件实现样本的后处理工作，如：选择是否打乱同组内数据，以及设定每组数据的最大长度。
+
+.. code-block:: python
+    
+    from ding.data.buffer.middleware import group_sample
+
+    buffer = DequeBuffer(size=10)
+    buffer.use(group_sample())
+
+
```
Comment (PaParaZz1) [inline source/04_best_practice/buffer_zh.rst:106]: 再介绍下两个东西吧：
- use_count middleware
- load data when initialization（用于模仿学习，比如dqfd之类的）

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**